### PR TITLE
Line2NodeMaterial: Add `opacityNode` support

### PR DESCRIPTION
**Pull request by Makio64**
Related issue: #[29962](https://github.com/mrdoob/three.js/issues/29962)

Add opacityNode support to line2NodeMaterial

wip : worked almost straightforward but at the intersection of line there is a round corner added
Screenshot 2024-11-25 at 23 13 24

### CODE REVIEW

**Comment by Makio64** [comment]
Screenshot 2024-11-25 at 23 23 47

It looks better but we can still see a point between segmentation

**Comment by Makio64** [comment]
I had to modify `LineGeometry` to fix the fact the segment were over each others..

also updated the demo 
Screenshot 2024-11-26 at 01 53 51

@Mugen87 let me know what you think of it, i didn't see bad side effects

File: src/materials/nodes/Line2NodeMaterial.js [hunk 1]
```diff
@@ -309,7 +310,7 @@ class Line2NodeMaterial extends NodeMaterial {
 					if ( useAlphaToCoverage && renderer.samples > 1 ) {
 
 						const dnorm = norm.fwidth();
-						alpha.assign( smoothstep( dnorm.negate().add( 0.5 ), dnorm.add( 0.5 ), norm ).oneMinus() );
+						alpha.assign( alpha.mul( smoothstep( dnorm.negate().add( 0.5 ), dnorm.add( 0.5 ), norm ).oneMinus() ) );
```
Comment (sunag) [inline src/materials/nodes/Line2NodeMaterial.js:313]: Let's try `alpha.mulAssign()` instead of `alpha.assign( alpha.mul(`

File: src/materials/nodes/Line2NodeMaterial.js [hunk 2]
```diff
@@ -40,6 +40,7 @@ class Line2NodeMaterial extends NodeMaterial {
 		this.lineWidth = 1;
 
 		this.lineColorNode = null;
+		this.opacityNode = null;
```
Comment (sunag) [inline src/materials/nodes/Line2NodeMaterial.js:43]: `opacityNode=null` seems unnecessary here, it is declared in `NodeMaterial` class.

**Comment by Mugen87** [comment]
Wide line materials do not support alpha because of the resulting artifacts, see #23680.

I'm not familiar enough with the wide line implementation so I hope @WestLangley or @gkjohnson can have a look at this PR.

**Comment by WestLangley** [comment]
Please test with wider lines and fewer points so you can see the artifacts. The segmented colors look like a previous issue.

Screenshot 2024-11-25 at 3 12 11 PM


Setting `transparent` to `true`, it looks pretty much the same for me. ???
Screenshot 2024-11-25 at 3 18 52 PM

File: examples/webgpu_lines_fat.html [hunk 3]
```diff
@@ -114,6 +114,8 @@
 
 				} );
 
+				matLine.opacityNode = uniform( float( 0.5 ) );
```
Comment (WestLangley) [inline examples/webgpu_lines_fat.html:117]: `float( 1 )`,  to match GIU.

**Comment by sunag** [comment]
You tried to multiply the opacity at the end instead of initially, and thus apply it to the entire shape?

**Comment by gkjohnson** [comment]
> Setting transparent to true, it looks pretty much the same for me. ???

Without something like a stencil test (see https://github.com/mrdoob/three.js/issues/23680#issuecomment-1063294691) it won't be possible to avoid these artifacts. Even with the pre-depth / stencil test, though, overlapping lines won't display correctly which would be even more noticeable with per-line opacity.

I don't have strong opinions on this other than it won't look right or work "out of the box" in the connected line segments case - this can be useful for users who know how to work around it, though. Separated line segments should look okay, as well. With more control over render order or ability to render an object with two passes in three.js it would be more possible to make a variant of this that "just works" (overlapping aside).

I wonder if something like a compute shader rasterizer or some custom per-segment depth adjustment could help with the artifacts.

File: examples/jsm/lines/LineGeometry.js [hunk 4]
```diff
@@ -15,19 +15,13 @@ class LineGeometry extends LineSegmentsGeometry {
...
-
-			points[ 2 * i ] = array[ i ];
-			points[ 2 * i + 1 ] = array[ i + 1 ];
-			points[ 2 * i + 2 ] = array[ i + 2 ];
-
-			points[ 2 * i + 3 ] = array[ i + 3 ];
-			points[ 2 * i + 4 ] = array[ i + 4 ];
-			points[ 2 * i + 5 ] = array[ i + 5 ];
+			points.set( array.slice( 3 * i, 3 * ( i + 1 ) ), 6 * i );
+			points.set( array.slice( 3 * ( i + 1 ), 3 * ( i + 2 ) ), 6 * i + 3 );
```
Comment (gkjohnson) [inline examples/jsm/lines/LineGeometry.js:24]: I'd prefer to revert these changes and keep the PR focused on any changes related to opacity. I find it harder to read and it will be slower and generate more garbage with the unnecessary "slice" calls.

File: src/materials/nodes/Line2NodeMaterial.js (hunk 1 above)
Comment (Makio64) [reply src/materials/nodes/Line2NodeMaterial.js]: changed to mulAssign

File: src/materials/nodes/Line2NodeMaterial.js (hunk 2 above)
Comment (Makio64) [reply src/materials/nodes/Line2NodeMaterial.js]: removed

File: examples/webgpu_lines_fat.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (Makio64) [reply examples/webgpu_lines_fat.html]: typo fixed, thanks!

File: examples/jsm/lines/LineGeometry.js (hunk 4 above)
Comment (Makio64) [reply examples/jsm/lines/LineGeometry.js]: revert to previous synthax

Comment (gkjohnson) [reply examples/jsm/lines/LineGeometry.js]: It looks like there's still a pretty significant diff in the LineGeometry.js file. I don't think it needs to be touched at all for the changes being proposed.

**Comment by Makio64** [comment]
@WestLangley I just tested with wider lines and it looks good on my `mac m3` see this screenshot : 

without alpha 
Screenshot 2024-11-26 at 12 19 58

with alpha 
Screenshot 2024-11-26 at 12 19 55

You need to apply my changes to the LineGeometry as well to make it works

But the problem is you can't see behind the line ( make sense as its draw as one object ? ) and more complex approach like suggest by @gkjohnson might be necessary for it ? Still I think its a little improvement compare to before and can be used to varius effect like making a line appear using alpha ( simulating line drawing as example ) or adding painting/particles textures to the line.

File: examples/jsm/lines/LineGeometry.js (hunk 4 above)
Comment (Makio64) [reply examples/jsm/lines/LineGeometry.js]: completly reverted

**Comment by Makio64** [comment]
I ported  https://github.com/pmndrs/meshline/blob/master/src/MeshLineGeometry.ts to nodeMaterial today and it's working great with alpha / uv / map / dash / gradient / etc.., i can clean it and will open a new PR.

Is there a reason why threejs used another approach than this one ?
If i name it MeshLineNodeMaterial / MeshLineGeometry will it be confusing ?

**Comment by Mugen87** [comment]
> Is there a reason why threejs used another approach than this one ?

This is years ago but originally we wanted to port this implementation, see #11040. We decided for #11349 though because it was more robust (`MeshLine` had issues with small point sets) and it was also more optimized because it allocates less memory on the GPU. I don't know if the current MeshLine` implementation has improved in that regard.
